src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs

```diff
@@ -456,13 +449,6 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Cancellat
                 {
                     var diagnostics = diagnosticsMap[analyzerWithState.Analyzer];
                     builder.AddRange(diagnostics.Where(ShouldInclude));
-
-                    // Save the computed diagnostics if caching is enabled and diagnostics were computed for the entire document.
-                    if (_cacheFullDocumentDiagnostics && !span.HasValue)
```

Comment:
Note this was purely a performance improvement added for LSP pull diagnostics as path as all code paths computing full document diagnostics (tagger, squiggles, etc.) were going through this path and were recomputing the same set of document diagnostics.

Reply:
We avoided doing this caching for non-LSP path as diagnostic incremental analyzer executed from solution crawler has logic to push diagnostics to error list from background analysis path, and if we ended caching diagnostics from other path, such as tagger, error list wasn't getting properly refreshed. So, it felt easier to just do this `_cacheFullDocumentDiagnostics` logic here for LSP path.

Reply:
> So if we were to add this back, i'd far prefer this caching just be in the LSP layer

That sounds reasonable to me.

> We need to start with correctness first, then work on optimizations.

I agree. Though I am still not able to follow the actual correctness issue here. We are caching full document diagnostics, just like we'd do in solution crawler path, and we don't have any known issues in the latter pertaining to caching full document diagnostics. One difference though is solution crawler executes callbacks into diagnostic incremental analyzer in a serial fashion, while the code path being removed in this PR can be executed concurrently from multiple clients.

> Note: this is the only issue currently blocking being able to enable pull diagnostics, so i'd like to make progress on this asap. Thanks!

Given this is just a performance aspect, I'd be fine to move ahead with a tracking issue filed to add this back. I can look next week and investigate this further.

Reply:
> Though I am still not able to follow the actual correctness issue here

Me neither. But it's definitely there :-)